File: tests/property/python/areas/quaternion_algebras.py
"""Live-SageMath oracle for quaternion algebras over QQ."""

from sage.all import *


def _vals(xs):
    return ','.join(str(x) for x in xs)


def quat_element(a, b, coeffs):
    A = QuaternionAlgebra(QQ, ZZ(a), ZZ(b), names=('i', 'j', 'k'))
    x = A([QQ(c) for c in coeffs])
    inv = '-' if x == 0 else _vals(list(~x))
    charpoly = _vals(x.reduced_characteristic_polynomial().list())
    return 'x=%s conj=%s tr=%s norm=%s inv=%s cp=%s' % (
        _vals(list(x)),
        _vals(list(x.conjugate())),
        x.reduced_trace(),
        x.reduced_norm(),
        inv,
        charpoly,
    )


def quat_product(a, b, left, right):
    A = QuaternionAlgebra(QQ, ZZ(a), ZZ(b), names=('i', 'j', 'k'))
    x = A([QQ(c) for c in left])
    y = A([QQ(c) for c in right])
    z = x * y
    return 'xy=%s pair=%s norm=%s anti=%s' % (
        _vals(list(z)),
        x.pair(y),
        z.reduced_norm(),
        (x * y).conjugate() == y.conjugate() * x.conjugate(),
    )


def quat_algebra(a, b):
    A = QuaternionAlgebra(QQ, ZZ(a), ZZ(b), names=('i', 'j', 'k'))
    return 'inv=%s disc=%s ram=%s definite=%s' % (
        _vals(A.invariants()),
        A.discriminant(),
        _vals(A.ramified_primes()),
        A.invariants()[0] < 0 and A.invariants()[1] < 0,
    )


FUNCTIONS = {
    'quat_element': quat_element,
    'quat_product': quat_product,
    'quat_algebra': quat_algebra,
}

File: tests/property/typescript/areas/quaternion_algebras.ts
/**
 * sagemath-ts side of the live quaternion-algebra differential area.
 */

import { QuaternionAlgebra } from '../../../../packages/sagemath-ts/src/algebras/quatalg/index.js';

type Printable = { toString(): string };
const vals = (xs: Printable[]): string => xs.map(String).join(',');
const bool = (value: boolean): string => (value ? 'True' : 'False');

function quat_element(a: bigint, b: bigint, coeffs: bigint[]): string {
  const A = QuaternionAlgebra(a, b);
  const x = A.__call__(coeffs);
  const inv = x.is_zero() ? '-' : vals(x.inverse().list());
  const charpoly = vals(x.reduced_characteristic_polynomial().coeffs);
  return `x=${vals(x.list())} conj=${vals(x.conjugate().list())} tr=${x.reduced_trace()} norm=${x.reduced_norm()} inv=${inv} cp=${charpoly}`;
}

function quat_product(a: bigint, b: bigint, left: bigint[], right: bigint[]): string {
  const A = QuaternionAlgebra(a, b);
  const x = A.__call__(left);
  const y = A.__call__(right);
  const z = x.mul(y);
  return `xy=${vals(z.list())} pair=${x.pair(y)} norm=${z.reduced_norm()} anti=${bool(
    z.conjugate().eq(y.conjugate().mul(x.conjugate()))
  )}`;
}

function quat_algebra(a: bigint, b: bigint): string {
  const A = QuaternionAlgebra(a, b);
  return `inv=${vals(A.invariants())} disc=${A.discriminant()} ram=${A.ramified_primes().join(
    ','
  )} definite=${bool(A.is_definite())}`;
}

export const functions = {
  quat_element,
  quat_product,
  quat_algebra,
};
